dali/imgcodec/image_decoder.cc

```diff
@@ -0,0 +1,614 @@
+                                  DecodeParams opts,
+                                  const ROI &roi) {
+  return ScheduleDecode(ctx, out, in, opts, roi).get_one(0);
+}
+
+
+/**
+ * @brief Decodes a single image to device buffers
+ */
+std::vector<DecodeResult> ImageDecoder::Decode(DecodeContext ctx,
+                                               span<SampleView<CPUBackend>> out,
+                                               cspan<ImageSource *> in,
+                                               DecodeParams opts,
+                                               cspan<ROI> rois) {
+  return ScheduleDecode(ctx, out, in, opts, rois).get_all_copy();
+}
+
+
+/**
+ * @brief Decodes a single image to device buffers
+ */
+std::vector<DecodeResult> ImageDecoder::Decode(DecodeContext ctx,
+                                               span<SampleView<GPUBackend>> out,
+                                               cspan<ImageSource *> in,
+                                               DecodeParams opts,
+                                               cspan<ROI> rois) {
+  return ScheduleDecode(ctx, out, in, opts, rois).get_all_copy();
+}
+
+
+
+FutureDecodeResults ImageDecoder::ScheduleDecode(DecodeContext ctx,
+                                                 SampleView<CPUBackend> out,
+                                                 ImageSource *in,
+                                                 DecodeParams opts,
+                                                 const ROI &roi) {
+  return ScheduleDecode(ctx, make_span(&out, 1), make_span(&in, 1), opts, make_span(&roi, 1));
+}
+
+FutureDecodeResults ImageDecoder::ScheduleDecode(DecodeContext ctx,
+                                                 SampleView<GPUBackend> out,
+                                                 ImageSource *in,
+                                                 DecodeParams opts,
+                                                 const ROI &roi) {
+  return ScheduleDecode(ctx, make_span(&out, 1), make_span(&in, 1), opts, make_span(&roi, 1));
+}
+
+FutureDecodeResults ImageDecoder::ScheduleDecode(DecodeContext ctx,
+                                                span<SampleView<CPUBackend>> out,
+                                                cspan<ImageSource *> in,
+                                                DecodeParams opts,
+                                                cspan<ROI> rois) {
+  temp_buffers_.clear();
+  int N = out.size();
+  assert(in.size() == N);
+  assert(rois.size() == N || rois.empty());
+  DecodeResultsPromise results(N);
+
+  auto work = new_work(ctx, results);
+  work->cpu_outputs.reserve(N);
+  work->indices.reserve(N);
+  work->sources.reserve(N);
+
+  for (auto &o : out)
+    work->cpu_outputs.push_back(o);
+  for (auto *src : in)
+    work->sources.push_back(src);
+  for (int i = 0; i < N; i++)
+    work->indices.push_back(i);
+
+  DistributeWork(std::move(work));
+
+  temp_buffers_.clear();
```

Comment:
I am struggling to understand why you need to clear again here. Can you explain?

Reply:
Moved to ScheduledWork.